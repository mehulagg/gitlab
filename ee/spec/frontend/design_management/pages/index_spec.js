import { createLocalVue, shallowMount } from '@vue/test-utils';
import createFlash from '~/flash';
import VueRouter from 'vue-router';
import Index from 'ee/design_management/pages/index.vue';
import uploadDesignQuery from 'ee/design_management/graphql/mutations/uploadDesign.mutation.graphql';
import DesignDestroyer from 'ee/design_management/components/design_destroyer.vue';

const localVue = createLocalVue();
localVue.use(VueRouter);
const router = new VueRouter({
  routes: [
    {
      name: 'designs',
      path: '/designs',
      component: Index,
    },
  ],
});

jest.mock('~/flash.js');

const mockDesigns = [
  {
    id: 'design-1',
    image: 'design-1-image',
    filename: 'design-1-name',
    event: 'NONE',
    notesCount: 0,
  },
  {
    id: 'design-2',
    image: 'design-2-image',
    filename: 'design-2-name',
    event: 'NONE',
    notesCount: 1,
  },
  {
    id: 'design-3',
    image: 'design-3-image',
    filename: 'design-3-name',
    event: 'NONE',
    notesCount: 0,
  },
];

const mockVersion = {
  node: {
    id: 'gid://gitlab/DesignManagement::Version/1',
  },
};

describe('Design management index page', () => {
  let mutate;
  let wrapper;

  const findDesignCheckboxes = () => wrapper.findAll('.design-checkbox');
  const findSelectAllButton = () => wrapper.find('.js-select-all');
  const findDeleteButton = () => wrapper.find('deletebutton-stub');

  function createComponent({
    loading = false,
    designs = [],
    allVersions = [],
    createDesign = true,
  } = {}) {
    mutate = jest.fn(() => Promise.resolve());
    const $apollo = {
      queries: {
        designs: {
          loading,
        },
        permissions: {
          loading,
        },
      },
      mutate,
    };

    wrapper = shallowMount(Index, {
      sync: false,
      mocks: { $apollo },
      localVue,
      router,
      stubs: { DesignDestroyer },
    });

    wrapper.setData({
      designs,
      allVersions,
      issueIid: '1',
      permissions: {
        createDesign,
      },
    });
  }

  afterEach(() => {
    wrapper.destroy();
  });

  describe('designs', () => {
    it('renders loading icon', () => {
      createComponent({ loading: true });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders error', () => {
      createComponent();

      wrapper.setData({ error: true });

      return wrapper.vm.$nextTick().then(() => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    it('renders empty text', () => {
      createComponent();

      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders designs list and header with upload button', () => {
      createComponent({ designs: mockDesigns, allVersions: [mockVersion] });

      return wrapper.vm.$nextTick().then(() => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    it('does not render toolbar when there is no permission', () => {
      createComponent({ designs: mockDesigns, allVersions: [mockVersion], createDesign: false });

      return wrapper.vm.$nextTick().then(() => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('onUploadDesign', () => {
    it('calls apollo mutate', () => {
      createComponent();

      return wrapper.vm
        .onUploadDesign([
          {
            name: 'test',
          },
        ])
        .then(() => {
          expect(mutate).toHaveBeenCalledWith({
            context: {
              hasUpload: true,
            },
            mutation: uploadDesignQuery,
            variables: {
              files: [{ name: 'test' }],
              projectPath: '',
              iid: '1',
            },
            update: expect.anything(),
            optimisticResponse: {
              __typename: 'Mutation',
              designManagementUpload: {
                __typename: 'DesignManagementUploadPayload',
                designs: [
                  {
                    __typename: 'Design',
                    id: expect.anything(),
                    image: '',
                    filename: 'test',
                    fullPath: '',
                    event: 'NONE',
                    notesCount: 0,
                    diffRefs: {
                      __typename: 'DiffRefs',
                      baseSha: '',
                      startSha: '',
                      headSha: '',
                    },
                    discussions: {
                      __typename: 'DesignDiscussion',
                      edges: [],
                    },
                    versions: {
                      __typename: 'DesignVersionConnection',
                      edges: {
                        __typename: 'DesignVersionEdge',
                        node: {
                          __typename: 'DesignVersion',
                          id: expect.anything(),
                          sha: expect.anything(),
                        },
                      },
                    },
                  },
                ],
              },
            },
          });
        });
    });

    it('does not call apollo mutate if createDesign is false', () => {
      createComponent({ createDesign: false });

      wrapper.vm.onUploadDesign([]);

      expect(mutate).not.toHaveBeenCalled();
    });

    it('sets isSaving', () => {
      createComponent();

      const uploadDesign = wrapper.vm.onUploadDesign([
        {
          name: 'test',
        },
      ]);

      expect(wrapper.vm.isSaving).toBe(true);

      return uploadDesign.then(() => {
        expect(wrapper.vm.isSaving).toBe(false);
      });
    });

    describe('upload count limit', () => {
      const MAXIMUM_FILE_UPLOAD_LIMIT = 10;

      afterEach(() => {
        createFlash.mockReset();
      });

      it('does not warn when the max files are uploaded', () => {
        createComponent();

        wrapper.vm.onUploadDesign(new Array(MAXIMUM_FILE_UPLOAD_LIMIT).fill(mockDesigns[0]));

        expect(createFlash).not.toHaveBeenCalled();
      });

      it('warns when too many files are uploaded', () => {
        createComponent();

        wrapper.vm.onUploadDesign(new Array(MAXIMUM_FILE_UPLOAD_LIMIT + 1).fill(mockDesigns[0]));

        expect(createFlash).toHaveBeenCalled();
      });
    });
  });

  describe('on latest version', () => {
    beforeEach(() => {
      createComponent({ designs: mockDesigns, allVersions: [mockVersion] });
    });

    it('renders design checkboxes', () => {
      expect(findDesignCheckboxes().length).toBe(mockDesigns.length);
    });

    it('renders Delete selected button', () => {
      expect(findDeleteButton().exists()).toBe(true);
    });

    it('renders a button with Select all text', () => {
      expect(findSelectAllButton().exists()).toBe(true);
      expect(findSelectAllButton().text()).toBe('Select all');
    });

    it('adds two designs to selected designs when their checkboxes are checked', () => {
      findDesignCheckboxes()
        .at(0)
        .trigger('click');
      findDesignCheckboxes()
        .at(1)
        .trigger('click');

      return wrapper.vm.$nextTick().then(() => {
        expect(findDeleteButton().exists()).toBe(true);
        expect(findSelectAllButton().text()).toBe('Deselect all');
        findDeleteButton().vm.$emit('deleteSelectedDesigns');
        const [{ variables }] = mutate.mock.calls[0];
        expect(variables.filenames).toStrictEqual([
          mockDesigns[0].filename,
          mockDesigns[1].filename,
        ]);
      });
    });

    it('adds all designs to selected designs when Select All button is clicked', () => {
      findSelectAllButton().vm.$emit('click');

      return wrapper.vm.$nextTick().then(() => {
        expect(findDeleteButton().props().hasSelectedDesigns).toBe(true);
        expect(findSelectAllButton().text()).toBe('Deselect all');
        expect(wrapper.vm.selectedDesigns).toEqual(mockDesigns.map(design => design.filename));
      });
    });

    it('removes all designs from selected designs when at least one design was selected', () => {
      findDesignCheckboxes()
        .at(0)
        .trigger('click');
      findSelectAllButton().vm.$emit('click');

      return wrapper.vm.$nextTick().then(() => {
        expect(findDeleteButton().props().hasSelectedDesigns).toBe(false);
        expect(findSelectAllButton().text()).toBe('Select all');
        expect(wrapper.vm.selectedDesigns).toEqual([]);
      });
    });
  });

  describe('on non-latest version', () => {
    beforeEach(() => {
      createComponent({ designs: mockDesigns, allVersions: [mockVersion] });

      router.replace({
        name: 'designs',
        query: {
          version: '2',
        },
      });
    });

    it('does not render design checkboxes', () => {
      expect(findDesignCheckboxes().length).toBe(0);
    });

    it('does not render Delete selected button', () => {
      expect(findDeleteButton().exists()).toBe(false);
    });

    it('does not render Select All button', () => {
      expect(findSelectAllButton().exists()).toBe(false);
    });
  });
});
