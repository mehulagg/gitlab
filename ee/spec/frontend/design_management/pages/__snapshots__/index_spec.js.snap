// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design management index page designs does not render toolbar when there is no permission 1`] = `
<div>
  <!---->
   
  <div
    class="mt-4"
  >
    <ol
      class="list-unstyled row"
    >
      <li
        class="col-md-6 col-lg-4 mb-3"
      >
        <design-stub
          event="NONE"
          filename="design-1-name"
          id="design-1"
          image="design-1-image"
          notescount="0"
        />
         
        <!---->
      </li>
      <li
        class="col-md-6 col-lg-4 mb-3"
      >
        <design-stub
          event="NONE"
          filename="design-2-name"
          id="design-2"
          image="design-2-image"
          notescount="1"
        />
         
        <!---->
      </li>
      <li
        class="col-md-6 col-lg-4 mb-3"
      >
        <design-stub
          event="NONE"
          filename="design-3-name"
          id="design-3"
          image="design-3-image"
          notescount="0"
        />
         
        <!---->
      </li>
    </ol>
  </div>
   
  <routerview-stub
    name="default"
  />
</div>
`;

exports[`Design management index page designs renders designs list and header with upload button 1`] = `
<div>
  <header
    class="row-content-block border-top-0 p-2 d-flex"
  >
    <div
      class="d-flex justify-content-between align-items-center w-100"
    >
      <designversiondropdown-stub />
       
      <div
        class="d-flex qa-selector-toolbar"
      >
        <glbutton-stub
          class="mr-2 js-select-all"
          variant="link"
        >
          Select all
        </glbutton-stub>
         
        <div>
          <deletebutton-stub
            buttonclass="btn-danger btn-inverted mr-2"
            buttonvariant=""
          >
            
            Delete selected
            
            <!---->
          </deletebutton-stub>
        </div>
         
        <uploadbutton-stub />
      </div>
    </div>
  </header>
   
  <div
    class="mt-4"
  >
    <ol
      class="list-unstyled row"
    >
      <li
        class="col-md-6 col-lg-4 mb-3"
      >
        <design-stub
          event="NONE"
          filename="design-1-name"
          id="design-1"
          image="design-1-image"
          notescount="0"
        />
         
        <input
          class="design-checkbox"
          type="checkbox"
        />
      </li>
      <li
        class="col-md-6 col-lg-4 mb-3"
      >
        <design-stub
          event="NONE"
          filename="design-2-name"
          id="design-2"
          image="design-2-image"
          notescount="1"
        />
         
        <input
          class="design-checkbox"
          type="checkbox"
        />
      </li>
      <li
        class="col-md-6 col-lg-4 mb-3"
      >
        <design-stub
          event="NONE"
          filename="design-3-name"
          id="design-3"
          image="design-3-image"
          notescount="0"
        />
         
        <input
          class="design-checkbox"
          type="checkbox"
        />
      </li>
    </ol>
  </div>
   
  <routerview-stub
    name="default"
  />
</div>
`;

exports[`Design management index page designs renders error 1`] = `
<div>
  <!---->
   
  <div
    class="mt-4"
  >
    <div
      class="alert alert-danger"
    >
      
      An error occurred while loading designs. Please try again.
    
    </div>
  </div>
   
  <routerview-stub
    name="default"
  />
</div>
`;

exports[`Design management index page designs renders loading icon 1`] = `
<div>
  <!---->
   
  <div
    class="mt-4"
  >
    <glloadingicon-stub
      color="orange"
      label="Loading"
      size="md"
    />
  </div>
   
  <routerview-stub
    name="default"
  />
</div>
`;

exports[`Design management index page when has no designs renders empty text 1`] = `
<div>
  <!---->
   
  <div
    class="mt-4"
  >
    <glemptystate-stub
      description="Upload and view the latest designs for this issue. Consistent and easy to find, so everyone is up to date."
      title="The one place for your designs"
    />
  </div>
   
  <routerview-stub
    name="default"
  />
</div>
`;
